rfcs/0075-declarative-wrappers.md

```diff
@@ -0,0 +1,353 @@
+setup hook)? Declarative wrappers would allow _some_ users to use the wrapped
+binaries and others not to need this wrapping. Via an override or a NixOS
+config flag, without triggering a rebuild of HPLIP itself, these users would be
+easily satisfied.
+
+## Orchestrating wrapping hooks
+
+- [issue 78792](https://github.com/NixOS/nixpkgs/issues/78792)
+
+@worldofpeace you are correct. All of these setup-hooks are a mess. At least we
+have documented, (yet not totally implemented) this section of the manual
+https://nixos.org/nixpkgs/manual/#ssec-gnome-common-issues-double-wrapped
+
+Declarative wrappers will deprecate the usage of our shell based hooks and will
+wrap all executables automatically according to their needs, without requiring
+the contributor a lot of knowledge of the wrapping system. Also, double
+wrappings will become a problem of the past.
+
+- [issue 86369](https://github.com/NixOS/nixpkgs/issues/86369) 
+
+@ttuegel I get the sense [you support this idea of declarative
+wrappers](https://github.com/NixOS/nixpkgs/issues/86369#issuecomment-626732191).
+For anyone else interested in a summary, the issue is a bit complex, so once
+you'll read the design of this RFC, and see examples of what the POC
+implementation of declarative wrappers [is capable
+of](https://github.com/NixOS/nixpkgs/pull/85103#issuecomment-614195666), I hope
+you'll see how declarative wrappers will solve this issue.
+
+
+## Issues _possibly_ fixable by declarative wrappers (?)
+
+- [pull 61213](https://github.com/NixOS/nixpkgs/pull/61213) 
+
+I'm not sure what's the issue there. But, I'm sure that a declarative, Nix
+based builder of a Python environment, even if this environment is used only
+for a build, should make it easier to control and alter it's e.g `$PATH`.
+
+- [issue 83667](https://github.com/NixOS/nixpkgs/issues/83667)
+
+@FRidh I see no reason for Python deps of Python packages to need to be in
+`propagatedBuildInputs` and not regular `buildInputs` but please correct me if
+I'm wrong. I think this was done so in the past so it'd be easy to know how to
+wrap them? Declarative wrappers won't require runtime-env-requiring deps to be
+only in `propagatedBuildInputs` or `buildInputs` - it should pick such deps
+from both lists. Hence, (I think) it should be possible to make Python's static
+builds consistent with other ecosystems.
+
+- [issue 86054](https://github.com/NixOS/nixpkgs/issues/86054)
+
+@ttuegel TBH I can't tell if declarative wrappers might help, but I'm linking
+this issue here because @worldofpeace wrote it might be related to wrappers?
+Feel free to suggest removing this in the RFC review.
+
+- [issue 49132](https://github.com/NixOS/nixpkgs/issues/49132)
+- [issue 54278](https://github.com/NixOS/nixpkgs/issues/54278)
+- [issue 39493](https://github.com/NixOS/nixpkgs/issues/39493)
+
+`GDK_PIXBUF` compatibilities? I haven't investigated them to the details, so feel
```

Comment:
Both of the issues related to gdk-pixbuf only happened for non-wrapped derivations and current `wrapGAppsHook` would already fix them.

The issue was, IIRC, that gdk-pixbuf NixOS module sets `GDK_PIXBUF_MODULE_FILE` environment variable, which was enabling binary gdk-pixbuf modules with ABI incompatible with the gdk-pixbuf the programs linked against. In [one case](https://github.com/NixOS/nixpkgs/issues/54278) it was caused by a bug that slipped through the ABI stability guarantees, in the [other](https://github.com/NixOS/nixpkgs/pull/60254) by running a program for `i686` on `x86_64`.

Either way, it is somewhat orthogonal to wrappers. Only that we might want to always wrap graphical applications with librsvg gdk-pixbuf module.